dali/c_api/c_api.cc

```diff
@@ -73,83 +131,99 @@ void daliPrefetchSeparate(daliPipelineHandle* pipe_handle,
   }
 }
 
-void daliSetExternalInput(daliPipelineHandle* pipe_handle, const char* name, device_type_t device,
-                          const void* data_ptr, dali_data_type_t data_type, const int64_t* shapes,
-                          int sample_dim, const char* layout_str) {
-  DALI_ENFORCE(device == CPU, "GPU data cannot be passed as external source");
-  dali::Pipeline* pipeline = reinterpret_cast<dali::Pipeline*>(pipe_handle->pipe);
-  std::vector<int64_t> shapes_tmp(shapes, shapes + sample_dim * pipeline->batch_size());
-  dali::TensorListShape<> tl_shape(std::move(shapes_tmp), pipeline->batch_size(), sample_dim);
-  dali::TensorLayout layout{};
-  if (layout_str != nullptr) {
-    layout = dali::TensorLayout(layout_str);
-  }
-  dali::TensorList<dali::CPUBackend> data;
-  const auto& type_info = dali::TypeTable::GetTypeInfo(static_cast<dali::DALIDataType>(data_type));
-  auto elem_sizeof = type_info.size();
-  // We cast away the const from data_ptr, as there is no other way of passing it to the
-  // TensorList, as we must also set the shape and type metadata.
-  // It is passed further as const TensorList, so it's data cannot be modified.
-  data.ShareData(const_cast<void*>(data_ptr), tl_shape.num_elements() * elem_sizeof);
-  data.Resize(tl_shape, type_info);
-  data.SetLayout(layout);
-  const auto& cdata = data;
-  pipeline->SetExternalInput(name, cdata);
+
+void daliSetExternalInput(daliPipelineHandle *pipe_handle, const char *name, device_type_t device,
+                          const void *data_ptr, dali_data_type_t data_type, const int64_t *shapes,
+                          int sample_dim, const char *layout_str) {
+  cudaStream_t stream;
+  CUDA_CALL(cudaStreamCreate(&stream));
+  daliSetExternalInputAsync(pipe_handle, name, device, data_ptr, data_type, shapes, sample_dim,
+                                 layout_str, stream);
+  CUDA_CALL(cudaStreamSynchronize(stream));
```

Comment:
You should call `cudaStreamDestroy`, or CUDAStream.
I don't know if creating a stream for every call isn't to time consuming operation.

Reply:
Maybe you should query DALI for a stream or create a static stream in this fucntion.

Reply:
Changed to create stream per pipeline